File: OS_kernel/py/10_intent_anchoring.py
"""
============================================================
KERNEL AGENTIX OS — 10_intent_anchoring.py
Module 10: Intent Anchoring
Penerjemah Niat Luar Menjadi Jangkar Topologi
============================================================
Depends on: corekernel.py
"""
from corekernel import AgentixKernel, Vector, Semantic


class IntentAnchoring(AgentixKernel):
    def __init__(self, config=None):
        super().__init__(config)
        self._anchors      = {}
        self._global_goals = {}

    def set_anchor(self, agent_id, intent_name, weight=1.0, target=None,
                   valence_target=None, decay=1.0, ttl=float('inf')):
        self._anchors.setdefault(agent_id, {})[intent_name] = {
            'weight':         max(0.0, min(1.0, weight)),
            'target':         list(target) if target else None,
            'valence_target': valence_target,
            'decay':          decay,
            'born':           self.iteration,
            'ttl':            ttl,
        }

    def clear_anchor(self, agent_id, intent_name):
        if agent_id in self._anchors:
            self._anchors[agent_id].pop(intent_name, None)

    def clear_all_anchors(self, agent_id):
        self._anchors[agent_id] = {}

    def resolve_anchors(self, agent_id):
        anchors = self._anchors.get(agent_id, {})
        now     = self.iteration
        active  = {}
        dead    = []
        for name, a in anchors.items():
            age      = now - a['born']
            is_alive = age < a['ttl']
            a['weight'] = max(0.0, a['weight'] * a['decay'])
            if a['weight'] > 1e-4 and is_alive:
                active[name] = a
            else:
                dead.append(name)
        for name in dead:
            del anchors[name]
        return active

    def intent_force(self, agent_id, agent_pos):
        net = Vector.zero()
        for a in self.resolve_anchors(agent_id).values():
            if not a or not a['target']: continue
            diff = Vector.sub(a['target'], agent_pos)
            dist = Vector.norm(diff) + 1e-12
            net  = Vector.add(net, Vector.mul(Vector.div(diff, dist), a['weight']))
        return net

    def valence_pull(self, agent_id, context):
        active  = self.resolve_anchors(agent_id)
        entries = [a for a in active.values() if a and a.get('valence_target')]
        if not entries: return 0.0
        pulls = [
            abs(context.get(a['valence_target']['field'], 0.0) - a['valence_target']['value']) * a['weight']
            for a in entries
        ]
        return min(1.0, sum(pulls) / len(pulls))

    def set_global_goal(self, goal_name, target, weight=0.5):
        self._global_goals[goal_name] = {'target': list(target), 'weight': weight}

    def global_intent_force(self, agent_pos):
        net = Vector.zero()
        for g in self._global_goals.values():
            diff = Vector.sub(g['target'], agent_pos)
            dist = Vector.norm(diff) + 1e-12
            net  = Vector.add(net, Vector.mul(Vector.div(diff, dist), g['weight']))
        return net

    def composite_force(self, agent_id, agent_pos, global_blend=0.3):
        local_f  = self.intent_force(agent_id, agent_pos)
        global_f = self.global_intent_force(agent_pos)
        return Vector.add(
            Vector.mul(local_f,  1.0 - global_blend),
            Vector.mul(global_f, global_blend),
        )

    def anchor_summary(self, agent_id):
        return [
            {'intent': name, 'weight': round(a['weight'], 3), 'has_target': bool(a['target'])}
            for name, a in self.resolve_anchors(agent_id).items()
        ]

File: OS_kernel/js/10_intent_anchoring.js
/**
 * ============================================================
 * KERNEL AGENTIX OS — 10_intent_anchoring.js
 * Module 10: Intent Anchoring
 * Penerjemah Niat Luar Menjadi Jangkar Topologi
 * ============================================================
 * Depends on: corekernel.js
 *
 * Intent Anchoring: translates high-level external intent
 *   (e.g. "score a goal", "defend zone", "follow leader")
 *   into persistent topological anchors — valence targets
 *   that continuously pull agent behavior.
 *
 * No if-then: intents are weighted attractors, not commands.
 * Multiple intents coexist and compete via RTD.
 */
import { AgentixKernel, Vector, Semantic } from './corekernel.js';

class IntentAnchoring extends AgentixKernel {
    constructor(config) {
        super(config);
        this._anchors    = {};   // { agentId: { intentName: AnchorDef } }
        this._globalGoals = {};  // system-wide shared goal anchors
    }

    // ─────────────────────────────────────────
    // ANCHOR DEFINITION
    // An anchor is an intent with:
    //   - weight: how strongly it pulls [0, 1]
    //   - target: optional spatial pos [x, y]
    //   - valenceTarget: desired context valence
    //   - decay: rate at which intent fades if not refreshed
    // ─────────────────────────────────────────
    setAnchor(agentId, intentName, {
        weight       = 1.0,
        target       = null,
        valenceTarget = null,
        decay        = 1.0,    // 1.0 = no decay, <1 = fades each tick
        ttl          = Infinity // time-to-live in ticks
    } = {}) {
        if (!this._anchors[agentId]) this._anchors[agentId] = {};
        this._anchors[agentId][intentName] = {
            weight:        Math.max(0, Math.min(1, weight)),
            target:        target ? [...target] : null,
            valenceTarget,
            decay,
            born:          this.iteration,
            ttl,
        };
    }

    clearAnchor(agentId, intentName) {
        if (this._anchors[agentId]) delete this._anchors[agentId][intentName];
    }

    clearAllAnchors(agentId) { this._anchors[agentId] = {}; }

    // ─────────────────────────────────────────
    // RESOLVE ANCHORS — get active anchors for agent
    // Expired (TTL) or decayed anchors are removed
    // ─────────────────────────────────────────
    resolveAnchors(agentId) {
        const anchors = this._anchors[agentId];
        if (!anchors) return {};

        const now = this.iteration;
        const active = {};

        Object.entries(anchors).forEach(([name, a]) => {
            const age       = now - a.born;
            const isAlive   = +(age < a.ttl);
            // Decay weight each tick
            a.weight        = Math.max(0, a.weight * (a.decay ** 1));
            const isValid   = +(a.weight > 1e-4) * isAlive;
            // Array dispatch: keep if valid, skip if not
            active[name]    = isValid ? a : undefined;
        });

        // Remove dead anchors
        Object.keys(active).forEach(name => {
            if (!active[name]) { delete anchors[name]; delete active[name]; }
        });

        return active;
    }

    // ─────────────────────────────────────────
    // INTENT FORCE VECTOR
    // Translates spatial anchors into attraction vector
    // Competes via RTD to produce single net direction
    // ─────────────────────────────────────────
    intentForce(agentId, agentPos) {
        const active = this.resolveAnchors(agentId);
        let net = Vector.zero();

        Object.values(active).forEach(a => {
            if (!a || !a.target) return;
            const diff  = Vector.sub(a.target, agentPos);
            const dist  = Vector.norm(diff) + 1e-12;
            const dir   = Vector.div(diff, dist);
            net = Vector.add(net, Vector.mul(dir, a.weight));
        });

        return net;
    }

    // ─────────────────────────────────────────
    // VALENCE PULL — how much does agent deviate
    // from all active valence targets?
    // Returns [0,1]: 0 = on target, 1 = fully off
    // ─────────────────────────────────────────
    valencePull(agentId, context) {
        const active = this.resolveAnchors(agentId);
        const entries = Object.values(active).filter(a => a && a.valenceTarget !== null);
        if (entries.length === 0) return 0;

        const pulls = entries.map(a => {
            const field = a.valenceTarget.field;
            const tgt   = a.valenceTarget.value;
            const curr  = context[field] || 0;
            return Math.abs(curr - tgt) * a.weight;
        });

        return Math.min(1, pulls.reduce((s, v) => s + v, 0) / entries.length);
    }

    // ─────────────────────────────────────────
    // GLOBAL GOAL — system-wide shared anchor
    // All agents without a specific anchor inherit this
    // ─────────────────────────────────────────
    setGlobalGoal(goalName, target, weight = 0.5) {
        this._globalGoals[goalName] = { target: [...target], weight };
    }

    globalIntentForce(agentPos) {
        let net = Vector.zero();
        Object.values(this._globalGoals).forEach(g => {
            const diff  = Vector.sub(g.target, agentPos);
            const dist  = Vector.norm(diff) + 1e-12;
            const dir   = Vector.div(diff, dist);
            net = Vector.add(net, Vector.mul(dir, g.weight));
        });
        return net;
    }

    // ─────────────────────────────────────────
    // COMPOSITE FORCE — agent anchors + global goals
    // Blended by local weight dominance
    // ─────────────────────────────────────────
    compositeForce(agentId, agentPos, globalBlend = 0.3) {
        const localF  = this.intentForce(agentId, agentPos);
        const globalF = this.globalIntentForce(agentPos);
        return Vector.add(
            Vector.mul(localF,  1 - globalBlend),
            Vector.mul(globalF, globalBlend)
        );
    }

    // ─────────────────────────────────────────
    // ANCHOR SUMMARY — inspect active intents
    // ─────────────────────────────────────────
    anchorSummary(agentId) {
        const active = this.resolveAnchors(agentId);
        return Object.entries(active).map(([name, a]) => ({
            intent: name,
            weight: a?.weight?.toFixed(3),
            hasTarget: !!a?.target,
        }));
    }
}

export { IntentAnchoring };
